### PR TITLE
feat(ARCO-139): submit unmined parent txs in beef with callbacks

### PR DESCRIPTION
**Pull request by kuba-4chain**
When sending BEEF to Arc with unmined ancestors of a transaction, ARC will return a status of the last transaction in BEEF, but will submit and register callbacks for all unmined transactions in BEEF.

### CODE REVIEW

File: internal/beef/beef.go [hunk 1]
```diff
@@ -59,28 +59,28 @@ func CheckBeefFormat(txHex []byte) bool {
...
-		return nil, nil, nil, err
+		return nil, nil, err
 	}
 
 	decodedBeef := &BEEF{
 		BUMPs:        bumps,
 		Transactions: transactions,
 	}
 
-	return decodedBeef.GetLatestTx(), decodedBeef, remainingBytes, nil
```
Comment (arkadiuszos4chain) [inline internal/beef/beef.go:83]: <3

File: test/beef_test.go [hunk 2]
```diff
@@ -34,17 +34,27 @@ func TestBeef(t *testing.T) {
...
+			beef, middleTx, tx := prepareBeef(t, txID, hash, address, dstAddress, privateKey)
 
 			body := api.POSTTransactionJSONRequestBody{
 				RawTx: beef,
 			}
 
+			callbackReceivedChan := make(chan *api.TransactionStatus)
+			callbackErrChan := make(chan error)
+
+			callbackUrl, token, shutdown := startCallbackSrv(t, callbackReceivedChan, callbackErrChan, nil)
```
Comment (arkadiuszos4chain) [inline test/beef_test.go:49]: Please move `startCallbackSrv(...)` to `utilis.go`

**Review by arkadiuszos4chain — CHANGES_REQUESTED** [review]

File: pkg/api/handler/helpers.go [hunk 3]
```diff
@@ -123,3 +123,28 @@ func convertMerkleRootsRequest(beefMerkleRoots []beef.MerkleRootVerificationRequ
...
+func findLastStatus(lastTxId string, statuses []*metamorph.TransactionStatus) *metamorph.TransactionStatus {
+	for _, status := range statuses {
+		if status.TxID == lastTxId {
+			return status
+		}
+	}
+	return nil
+}
+
+func filterStatusesOfInterest(txIds []string, allStatuses []*metamorph.TransactionStatus) []*metamorph.TransactionStatus {
```
Comment (boecklim) [inline pkg/api/handler/helpers.go:136]: I'd rename this function to `filterStatusesByTxIDs(txIds []string, txStatuses []*metamorph.TransactionStatus)`

File: test/beef_test.go [hunk 4]
```diff
@@ -62,6 +72,32 @@ func TestBeef(t *testing.T) {
 			statusResponse, err := arcClient.GETTransactionStatusWithResponse(context.Background(), tx.TxID())
 			require.NoError(t, err)
 			require.Equal(t, metamorph_api.Status_MINED.String(), *statusResponse.JSON200.TxStatus)
+
+			// verify callbacks for both unmined txs in BEEF
+			lastTxCallbackReceived := false
+			middleTxCallbackReceived := false
+
+			for i := 0; i < 2; i++ {
```
Comment (boecklim) [inline test/beef_test.go:80]: Are 2 callbacks expected?
Maybe add a variable there like `expectedCallbacks := 2` to make it more explicit

File: pkg/api/handler/helpers.go [hunk 5]
```diff
@@ -123,3 +123,28 @@ func convertMerkleRootsRequest(beefMerkleRoots []beef.MerkleRootVerificationRequ
 
 	return merkleRoots
 }
+
+func findLastStatus(lastTxId string, statuses []*metamorph.TransactionStatus) *metamorph.TransactionStatus {
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply pkg/api/handler/helpers.go]: Good point, done

File: test/beef_test.go (hunk 2 above)
Comment (kuba-4chain) [reply test/beef_test.go:49]: Done

File: test/beef_test.go (hunk 4 above)
Comment (kuba-4chain) [reply test/beef_test.go]: Yes, 2 callbacks are expected, one for the last tx in beef and one for the unmined parent, as per this feature request.

Thanks for pointing that out, magic numbers are the worst. Fixed

**Review by boecklim — APPROVED** [review]

File: test/beef_test.go (hunk 4 above)
Comment (arkadiuszos4chain) [reply test/beef_test.go]: IMO, making this magic number a variable doesn't make the code more readable or clear; moreover, it is misleading.
1. Making it a variable lets other developers think it is configurable via the variable, which is not true.
2. The expected callback number is hardcoded in the prepareBeef() function.

To make it truly readable, we should:
1. Change the prepareBeef() function implementation to make it possible to configure how many transactions it should contain (mined and unmined) – IMO, overengineering.
2. Return the number of "unmined" transactions from prepareBeef() – this way, we are hiding all the mess from the test code and move it to the setup code.

**Pull request merged**